vm/devices/storage/disk_backend/src/layered.rs

```diff
@@ -0,0 +1,718 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! A layered disk implementation, [`LayeredDisk`].
+//!
+//! A layered disk is a disk composed of multiple layers. Each layer is a block
+//! device made up of sectors, but with the added per-sector state of whether
+//! the sector is present or not. When reading a sector, the layered disk will
+//! read from the topmost layer that has the sector present. When writing, the
+//! disk will write to the topmost layer.
+//!
+//! A layer can also have caching behavior. If a layer is configured to cache
+//! reads, then sectors that are read from lower layers are written back to the
+//! layer. If a layer is configured to write through, then writes are written to
+//! the layer and the next layer. These can be useful to implement simple
+//! persistent and non-persistent caches, primarily designed for lazily
+//! populating local backing stores from remote sources.
+//!
+//! Missing from this implementation is write-back caching and cache eviction,
+//! which would be needed for caches that are smaller than the disk. These
+//! require potentially complicated cache management policies and are probably
+//! best implemented in a separate disk implementation.
+
+mod bitmap;
+
+pub use bitmap::SectorMarker;
+
+use crate::Disk;
+use crate::DiskError;
+use crate::DiskIo;
+use crate::Unmap;
+use bitmap::Bitmap;
+use guestmem::MemoryWrite;
+use inspect::Inspect;
+use scsi_buffers::RequestBuffers;
+use std::future::Future;
+use std::pin::Pin;
+use thiserror::Error;
+
+/// A disk composed of multiple layers.
+#[derive(Inspect)]
+pub struct LayeredDisk {
+    #[inspect(iter_by_index)]
+    layers: Vec<Layer>,
+    is_read_only: bool,
+    is_fua_respected: bool,
+    sector_shift: u32,
+    disk_id: Option<[u8; 16]>,
+    physical_sector_size: u32,
+    optimal_unmap_sectors: Option<u32>,
+}
+
+#[derive(Inspect)]
+struct Layer {
+    backing: Box<dyn DynLayer>,
+    visible_sector_count: u64,
+    behavior: LayerBehavior,
+}
+
+/// The caching behavior of the layer.
+#[derive(Clone, Debug, Inspect, Default)]
```

Comment:
I'd swap the implicit `Default` with a explicit (and appropriately named) `new` method that makes it more obvious what the "Default" semantics actually are